src/cli_commands.h

```diff
@@ -0,0 +1,57 @@
+#ifndef __REDIS_CLI_COMMANDS_H
+#define __REDIS_CLI_COMMANDS_H
+
+#include <stddef.h>
+
+#define REDIS_CLI_COMMANDS
+
+typedef enum {
+    ARG_TYPE_STRING,
+    ARG_TYPE_INTEGER,
+    ARG_TYPE_DOUBLE,
+    ARG_TYPE_KEY, /* A string, but represents a keyname */
+    ARG_TYPE_PATTERN,
+    ARG_TYPE_UNIX_TIME,
+    ARG_TYPE_PURE_TOKEN,
+    ARG_TYPE_ONEOF, /* Has subargs */
+    ARG_TYPE_BLOCK /* Has subargs */
+} commandArgType;
```

Comment:
This enum could be extracted to a header file without a problem.

The args struct we could find a way to share, but there are fields used by client and not the server (the arg matching fields), and vice versa (key spec, deprecated, summary). 

The redis command struct cannot realistically be shared by both. The `redisCommand` struct is massive, and it pulls in many other data types, including the command function pointer, histogram, key specs, get key function pointer, the module command data, etc. Separating it from server.h is possible but very difficult and not likely to be clean.

Reply:
I don't have a strong opinion regarding where to define the macros.

Reply:
i must be missing something, but i don't understand how there are fields that are used by the cli and not by the server. after all, till now, anyone who used any info from the json files did it by taking it from COMMAND command (possibly converted into json).

i imagine that it could be that one day we'll have other tools in our repo that want to include this commands.c file, and they might need other bits of info.

it easy to filter out the proc and getkeys proc from the data (replacing it with NULL).
regarding the other excessive fields on redisCommand struct, which we don't want to pull into the header file, i suppose we should take this opportunity to split this struct into two parts, anything below the `Runtime populated data` comment should move into another struct, and then one struct should point to the other.
i.e. either the runtime struct will have a pointer to the static struct, or the static struct will have a `void*` member that's set to a heap allocated runtime struct (i prefer the first approach, will make a cleaner code).
another advantage of that would be smaller executable size, since we don't reserve the histogram and other fields in the data segment.

Reply:
Or we could leave a void* field for use as a runtime scratchpad by users of the arg struct.

Reply:
I still think splitting the redisCommand struct - as much as I agree that it needs to happen - is a much bigger job than you realize. It drags with it several other nontrivial structs and typedefs, some of which are themselves also used elsewhere. And all the code that accesses those fields will need to be modified, along with the code that populates them.

This will end up being a significant refactoring to tack on to this PR.

Reply:
If you want to do this, I'd suggest implementing the refactoring in a separate PR which does only that: move the command-related headers from `server.h` to `commands.h` and split `redisCommand` into two structs. The this PR can be rebased onto that.

Reply:
Ok, but neither of them is gonna be merged soon (we decided that considering such a big change, the feature isn't justifying the risk of breaking something in a patch level release). 
So I rather not wait, and keep progressing here. 
We can make another PR, and merge the same commits here. Or add the commits here, and later before merging, extract them to another PR.

Reply:
So for now I won't be splitting `redisCommand` - that's too big a change.

I haven't changed the args struct either - for now the server and cli have different structs due to the runtime data. If you still think it's important I'll take another look at that.

Reply:
i agree, let's cut our losses and keep separate structs.